### ISSUE TITLE
[Bug] Discrepancy in tf.keras and keras in setting model.trainable = False and then compiling

### ISSUE THREAD

**Issue by naoto0804** [tensorflow/tensorflow#22012]
### System information
- Have I written custom code: Yes
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from: binary
- TensorFlow version: v1.10.1-0-g4dcfddc5d1 1.10.1
- Bazel version: N/A
- CUDA/cuDNN version: CUDA9.1, cuDNN7.0
- GPU model and memory: TITAN V
- Exact command to reproduce: python3 compare.py
- Mobile device: N/A

### Describe the problem
This is a toy case for training GAN problem.
When I run the code shown below in keras2.2.2, I get
```
WARNING:tensorflow:Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?
```
only at the beginning of the training once.
However, I run this code in tensorflow1.10.1, the warning raises at every iteration.
Although it seems that the model is appropriately learned (I can make sure the weight is freezed by the result of `.summary()`), raising too many warning is not torelable.

Here is the complete log.
[tf.keras ver](https://github.com/tensorflow/tensorflow/files/2342979/tf_keras.log)
[keras ver](https://github.com/tensorflow/tensorflow/files/2342980/keras.log)

I see the same problem is posted in Stackoverflow
https://stackoverflow.com/questions/50468940/tensorflow-1-8-tf-keras-gives-different-result-in-dcgan-from-keras

```
import numpy as np

# use keras
from keras.layers import Dense, Input
from keras.models import Model

# use tf.keras
# from tensorflow.keras.layers import Dense, Input
# from tensorflow.keras.models import Model

# define input
noise = Input(shape=(10,))
x = Input(shape=(100,))

# define generator and discriminator
gen = Dense(100)
dis = Dense(1)

y = dis(x)
dis_model = Model(x, y)
dis_model.compile(optimizer='rmsprop', loss='mse')
dis_model.summary()

z = dis_model(gen(noise))
dis_model.trainable = False
combined_model = Model(noise, z)
combined_model.compile(optimizer='rmsprop', loss='mse')
combined_model.summary()

for i in range(3):
    dis_model.train_on_batch(x=np.random.rand(10, 100),
                             y=np.random.rand(10, 1))
    combined_model.train_on_batch(x=np.random.rand(10, 10),
                             y=np.random.rand(10, 1))

```

**Comment by yanboliang** [comment]
@naoto0804 
Keras model training will update params in ```model._collected_trainable_weights``` which is produced at ```model.compile```.
When you call ```dis_model.trainable = False``` to freeze ```dis_model```, you don't compile ```dis_model```, so two things happens:
* ```dis_model._collected_trainable_weights``` still includes old trainable params.
* ```dis_model.trainable_weights``` will return null, as you set all params non-trainable.

When you call ```dis_model.train_on_batch```, the params in ```dis_model._collected_trainable_weights``` will be updated according gradient descent, so it meets your requirements. But Keras found ```dis_model._collected_trainable_weights``` and ```dis_model.trainable_weights``` are inconsistent, so print warning incase this is not by design.

When you call ```combined_model.compile```, ```combined_model._collected_trainable_weights``` will be generated, and it doesn't includes the params in ```dis_model``` as they are marked as non-trainable.

To sum up, this warning is to prevent misuse. In your case, you code makes sense. I think we just need to reduce log frequency in ```tf.keras``` to once only, in case to flood the screen. I sent #22296 to fix it. Thanks.

**Comment by naoto0804** [comment]
I'm bit new to tf/keras. Thank you so much for clear explanation and PR!

**Comment by fchollet** [comment]
Note that you can also suppress the warning by setting back `dis_model.trainable = True` once you've compiled `combined_model`.

**Comment by naoto0804** [comment]
That's really straightforward way, thank you for the suggestion!

**Comment by lzhengchun** [comment]
> d `combined_mod`

this does not seem to work

**Comment by lzhengchun** [comment]
> Note that you can also suppress the warning by setting back `dis_model.trainable = True` once you've compiled `combined_model`.

+1


> @naoto0804
> Keras model training will update params in `model._collected_trainable_weights` which is produced at `model.compile`.
> When you call `dis_model.trainable = False` to freeze `dis_model`, you don't compile `dis_model`, so two things happens:
> 
> * `dis_model._collected_trainable_weights` still includes old trainable params.
> * `dis_model.trainable_weights` will return null, as you set all params non-trainable.
> 
> When you call `dis_model.train_on_batch`, the params in `dis_model._collected_trainable_weights` will be updated according gradient descent, so it meets your requirements. But Keras found `dis_model._collected_trainable_weights` and `dis_model.trainable_weights` are inconsistent, so print warning in case this is not by design.
> 
> When you call `combined_model.compile`, `combined_model._collected_trainable_weights` will be generated, and it doesn't includes the params in `dis_model` as they are marked as non-trainable.
> 
> To sum up, this warning is to prevent misuse. In your case, you code makes sense. I think we just need to reduce log frequency in `tf.keras` to once only, to prevent it floods the screen. I sent #22296 to fix it. Thanks.

after setting  `dis_model.trainable = True`, ` dis_model._collected_trainable_weights`, and `dis_model.trainable_weights` match but it still gives warning, puzzling.
